tests/h/accounts/schemas_test.py

```diff
@@ -377,27 +377,138 @@ def test_it_is_invalid_if_password_wrong(self,
-                                             factories,
-                                             pyramid_csrf_request,
-                                             user_model):
-        # There isn't already an account with the email address that we're
-        # trying to change to.
-        user_model.get_by_email.return_value = None
+    # The user's password.
+    PASSWORD = 'flibble'
+
+    def test_it_returns_the_new_email_when_valid(self, schema):
+        appstruct = schema.deserialize({
+            'email': 'foo@bar.com',
+            'password': self.PASSWORD,
+        })
+
+        assert appstruct['email'] == 'foo@bar.com'
+
+    def test_it_is_valid_if_email_same_as_users_existing_email(self,
+                                                               schema,
+                                                               user,
+                                                               user_model,
+                                                               pyramid_config):
+        """
+        It is valid if the new email is the same as the user's existing one.
+
+        Trying to change your email to what your email already is should not
+        return an error.
+
+        """
+        user_model.get_by_email.return_value = Mock(spec_set=['userid'],
+                                                    userid=user.userid)
+        pyramid_config.testing_securitypolicy(user.userid)
+
+        schema.deserialize({'email': user.email, 'password': self.PASSWORD})
+
+    def test_it_is_invalid_if_csrf_token_missing(self,
+                                                 pyramid_request,
+                                                 schema):
+        del pyramid_request.headers['X-CSRF-Token']
 
-        pyramid_csrf_request.authenticated_user = factories.User()
-        db_session.add(pyramid_csrf_request.authenticated_user)
+        with pytest.raises(BadCSRFToken):
+            schema.deserialize({
+                'email': 'foo@bar.com',
+                'password': self.PASSWORD,
+            })
 
-        schema = schemas.EmailChangeSchema().bind(request=pyramid_csrf_request)
+    def test_it_is_invalid_if_csrf_token_wrong(self, pyramid_request, schema):
+        pyramid_request.headers['X-CSRF-Token'] = 'WRONG'
 
+        with pytest.raises(BadCSRFToken):
+            schema.deserialize({
+                'email': 'foo@bar.com',
+                'password': self.PASSWORD,
+            })
+
+    def test_it_is_invalid_if_password_wrong(self, schema):
         with pytest.raises(colander.Invalid) as exc:
             schema.deserialize({
                 'email': 'foo@bar.com',
-                'password': 'flibble'  # Not the correct password!
+                'password': 'WRONG'  # Not the correct password!
             })
 
-        assert 'password' in exc.value.asdict()
+        assert exc.value.asdict() == {
+            'password': 'Incorrect password. Please try again.'}
```

Comment:
I realize that asserting the actual error messages is a little brittle, but it's not too bad (you change an error message, you have to update one test) and while writing these tests I did come across cases where `Invalid` would be raised but for some reason other than what the test was trying to test for (so the test was not actually testing what it thought it was) and where an error message would be raised on the right field but it would be the wrong error message. It used to be that if you tried to change your email address, and entered "a" in the password field, you could get a "password too short" error.